micropython/test_micropython.c

```diff
@@ -0,0 +1,157 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Used for MicroPython testing. Runs MicroPython script.
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Piotr Nieciecki
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <string.h>
+
+#define MICROPYTHON   "/bin/micropython "
+#define PATH_TO_TESTS "/usr/test/micropython/"
+
+#define CMDLINE_OPTION_IDEN "# cmdline: "
+#define OPTION_IDEN_LEN     strlen(CMDLINE_OPTION_IDEN)
+
+#define TRUE  1
+#define FALSE 0
+
+static char *PROG_NAME = "";
+
+void upyth_errMsg(char *msg)
+{
+	fprintf(stderr, "Error: %s - %s\n", PROG_NAME, msg);
+}
+
+char *upyth_concat(const char *str1, const char *str2)
+{
+	int len;
+	char *res;
+
+	len = strlen(str1) + strlen(str2) + 1;
+	res = (char *)malloc(sizeof(char) * len);
+
+	if (res != NULL) {
+		res[0] = '\0';
+		strcat(res, str1);
+		strcat(res, str2);
+	}
+	return res;
+}
+
+char *upyth_optionsGet(const char *path, int *Error)
+{
+	FILE *f;
+	char *line = NULL;
+	size_t len;
+	char *result = NULL;
+
+	f = fopen(path, "r");
+	if (f == NULL) {
+		upyth_errMsg("There was a problem with opening file with a script");
+		Error = TRUE;
+		return NULL;
+	}
+
+	if (getline(&line, &len, f) < 0) {
+		fclose(f);
+		upyth_errMsg("There was a problem with reading script to execute");
+		Error = TRUE;
+		return NULL;
+	}
+
+	len = strlen(line) + 1;
+
+	if (len > OPTION_IDEN_LEN && strncmp(line, CMDLINE_OPTION_IDEN, OPTION_IDEN_LEN) == 0) {
+		result = (char *)malloc(sizeof(char) * (strlen(line) - OPTION_IDEN_LEN));
+		if (result == NULL) {
+			fclose(f);
+			free(line);
+			upyth_errMsg("Malloc error");
+			Error = TRUE;
+			return NULL;
+		}
+
+		strcpy(result, &line[OPTION_IDEN_LEN]);
+
+		/* Changing the last character from '\n' to ' ' */
+		result[strlen(result) - 1] = ' ';
+	}
+
+	free(line);
+	fclose(f);
+
+	return result;
+}
+
+int main(int argc, char **argv)
+{
+	char *cmd, *options, *tmp;
+	int upythProgRes, err = FALSE;
+
+	PROG_NAME = argv[0];
+
+	if (argc != 2) {
+		upyth_errMsg("There is no required argument - please execute this function with path to the micropython tests");
+		return 1;
+	}
+
+	if (chdir(PATH_TO_TESTS) != 0) {
+		upyth_errMsg("There is no such a micropython test to run, build project with \"LONG_TEST=y\"");
+		return 1;
+	}
+
+	if (access(argv[1], F_OK) != 0) {
+		upyth_errMsg("There is no such a micropython test to run, build project with \"LONG_TEST=y\"");
+		return 1;
+	}
+
+	printf("Running test: %s%s\n", PATH_TO_TESTS, argv[1]);
+
+	tmp = MICROPYTHON;
```

Comment:
Maybe change `concat` to:
```c
char *upyth_concat(char *str1, const char *str2)
```
and use `realloc` inside to lower memory fragmentation.